lib/core/network/protocols/http+websocket.js

```diff
@@ -0,0 +1,920 @@
+    this.server.ws('/*', {
+      ...this.wsConfig.opts,
+      maxBackPressure: WS_MAX_BACKPRESSURE,
+      open: this.wsOnOpenHandler.bind(this),
+      close: this.wsOnCloseHandler.bind(this),
+      message: this.wsOnMessageHandler.bind(this),
+      drain: this.wsOnDrainHandler.bind(this),
+    });
+    /* eslint-enable sort-keys */
+  }
+
+  initHttp () {
+    this.server.any('/*', this.httpOnMessageHandler.bind(this));
+  }
+
+  broadcast (data) {
+    const stringified = JSON.stringify(data.payload);
+    const payloadByteSize = Buffer.from(stringified).byteLength;
+    // 255 bytes should be enough to hold the following:
+    //     ,"room":"<channel identifier>"
+    // (with current channel encoding, this is less than 100 bytes)
+    const payload = Buffer.allocUnsafe(payloadByteSize + 255);
+
+    let offset = payloadByteSize - 1;
+
+    payload.write(stringified, 0);
+    payload.write(JSON_ROOM_PROPERTY, offset);
+
+    offset += JSON_ROOM_PROPERTY.length;
+
+    for (const channel of data.channels) {
+      // Adds the room property to the message
+      payload.write(channel, offset);
+      payload.write(JSON_ENDER, offset + channel.length);
+
+      // prevent buffer overwrites due to sending packets being an
+      // async method (race condition)
+      const payloadLength = offset + channel.length + JSON_ENDER.length;
+      const payloadSafeCopy = Buffer.allocUnsafe(payloadLength);
+
+      payload.copy(payloadSafeCopy, 0, 0, payloadLength);
+
+      debugWS('Publishing to channel "realtime/%s": %s', channel, payloadSafeCopy);
+      this.server.publish(`realtime/${channel}`, payloadSafeCopy, false);
+    }
+  }
+
+  notify (data) {
+    const socket = this.socketByConnectionId.get(data.connectionId);
+    debugWS('notify: %a', data);
+
+    if (!socket) {
+      return;
+    }
+
+    const payload = data.payload;
+
+    for (let i = 0; i < data.channels.length; i++) {
+      payload.room = data.channels[i];
+      this.wsSend(socket, Buffer.from(JSON.stringify(payload)));
+    }
+  }
+
+  joinChannel (channel, connectionId) {
+    debugWS('joinChannel: %s %s', channel, connectionId);
+
+    const socket = this.socketByConnectionId.get(connectionId);
+
+    if (!socket) {
+      return;
+    }
+
+    debugWS('Subscribing connection ID "%s" to channel "realtime/%s"', connectionId, channel);
+    socket.subscribe(`realtime/${channel}`);
+  }
+
+  leaveChannel (channel, connectionId) {
+    debugWS('leaveChannel: %s %s', channel, connectionId);
+
+    const socket = this.socketByConnectionId.get(connectionId);
+
+    if (!socket) {
+      return;
+    }
+
+    socket.unsubscribe(`realtime/${channel}`);
+  }
+
+  disconnect (connectionId, message = null) {
+    debugWS('[%s] forced disconnect', connectionId);
+
+    const socket = this.socketByConnectionId.get(connectionId);
+
+    if (!socket) {
+      return;
+    }
+
+    socket.end(1011, message ? Buffer.from(message) : WS_GENERIC_CLOSE_MESSAGE);
```

Comment:
What is this number `1011` ?

Reply:
https://tools.ietf.org/html/rfc6455#page-46

Reply:
Can we have a named constant instead of a number? It's more readable than having to read the RFC :sweat_smile: